File: Structural/Decorator/Cpp/SugarDecorator.cpp
// SugarDecorator.cpp

#include "SugarDecorator.h"

SugarDecorator::SugarDecorator(std::unique_ptr<ICoffee> coffee)
  : CoffeeDecorator(std::move(coffee)) {}

std::string SugarDecorator::getDescription() const {
  return decoratedCoffee->getDescription() + ", Sugar";
}

double SugarDecorator::getCost() const {
  return decoratedCoffee->getCost() + 0.5;
}

File: Structural/Decorator/CSharp/SugarDecorator.cs
// SugarDecorator.cs

public class SugarDecorator : CoffeeDecorator {
  public SugarDecorator(ICoffee coffee) : base(coffee) {}

  public override string GetDescription() {
    return decoratedCoffee.GetDescription() + ", Sugar";
  }

  public override double GetCost() {
    return decoratedCoffee.GetCost() + 0.5;
  }
}
